_posts/2017-08-31-Rust-1.20.md

```diff
@@ -0,0 +1,261 @@
+}
+```
+
+That is, the constant `ID` is associated with `Struct`. Like functions,
+associated constants work with traits and enums as well.
+
+Traits have an extra ability with associated constants that gives them some
+extra power. With a trait, you can use an associated constant in the same way
+you'd use an associated type: by declaring it, but not giving it a value. The
+implementor of the trait then declares its value upon implementation:
+
+```rust
+trait Trait {
+    const ID: u32;
+}
+
+struct Struct;
+
+impl Trait for Struct {
+    const ID: u32 = 5;
+}
+
+fn main() {
+    println!("{}", Struct::ID);
+}
+```
+
+Before this release, if you wanted to make a trait that represented floating
+point numbers, you'd have to write this:
+
+```rust
+trait Float {
+    fn nan() -> Self;
+    fn infinity() -> Self;
+    ...
+}
+```
+
+This is slightly unweildy, but more importantly, becuase they're functions, they
+cannot be used in constant expressions, even though they only return a constant.
+Because of this, a design for `Float` would also have to include constants as well:
+
+```rust
+mod f32 {
+    const NAN: f32 = 0.0f32 / 0.0f32;
+    const INFINITY: f32 = 1.0f32 / 0.0f32;
+
+    impl Float for f32 {
+        fn nan() -> Self {
+            f32::NAN
+        }
+        fn infinity() -> Self {
+            f32::INFINITY
+        }
+    }
+}
+```
+
+Associated constants let you do this in a much cleaner way. This trait definition:
+
+```rust
+trait Float {
+    const NAN: Self;
+    const INFINITY: Self;
+    ...
+}
+```
+
+Leads to this implementation:
+
+```rust
+mod f32 {
+    impl Float for f32 {
+        const NAN: f32 = 0.0f32 / 0.0f32;
+        const INFINITY: f32 = 1.0f32 / 0.0f32;
+    }
+}
+```
+
+much cleaner, and more versitile.
+
+Associated constants were proposed in [RFC 195], almost exactly three years ago. It's
+been quite a while for this feature! That RFC contained all associated items, not just
+constants, and so some of them, such as associated types, were implemented faster than
+others. In general, we've been doing a lot of internal work for constant evaluation,
+to increase Rust's capabilities for compile-time metaprogramming. Expect more on this
+front in the future.
+
+[RFC 195]: https://github.com/rust-lang/rfcs/blob/master/text/0195-associated-items.md
+
+Back in [Rust 1.14], we announced preliminary support for asm.js and wasm
+with the help of Emscripten. Since then, LLVM has added its own support, and
+so, we have [added native wasm support to Rust]! It's using the
+`wasm32-experimental-emscripten` target.
+
+[Rust 1.14]: https://blog.rust-lang.org/2016/12/22/Rust-1.14.html
+[added native wasm support to Rust]: https://github.com/rust-lang/rust/pull/42571
+
+See the [detailed release notes][notes] for more.
+
+#### Library stabilizations
+
+There's nothing *super* exciting in libraries this release, just a number of solid
+improvements and continued stabilizing of APIs.
```

Comment:
I think this section should mention unstable sort, which is often faster than stable sort and actually works with just libcore too.

Reply:
That wasn't in the release notes, can you link me to a PR?

Reply:
It is in the release notes:
-----
The following APIs were also stabilized:
...
- [`slice::sort_unstable_by_key`], [`slice::sort_unstable_by`], and [`slice::sort_unstable`]

[`slice::sort_unstable_by_key`]: https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable_by_key
[`slice::sort_unstable_by`]: https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable_by
[`slice::sort_unstable`]: https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable

Reply:
ah sorry, i see now. yeah, sounds good!